Lib/ufo2ft/compilers/baseCompiler.py

```diff
@@ -0,0 +1,304 @@
+import logging
+import os
+from collections import defaultdict
+from dataclasses import dataclass
+from typing import Callable, Optional, Type
+
+from fontTools.designspaceLib.split import splitInterpolable, splitVariableFonts
+from fontTools.misc.loggingTools import Timer
+from fontTools.otlLib.optimize.gpos import GPOS_COMPACT_MODE_ENV_KEY
+
+from ufo2ft.constants import MTI_FEATURES_PREFIX
+from ufo2ft.errors import InvalidDesignSpaceData
+from ufo2ft.featureCompiler import FeatureCompiler, MtiFeatureCompiler
+from ufo2ft.postProcessor import PostProcessor
+from ufo2ft.util import (
+    _notdefGlyphFallback,
+    colrClipBoxQuantization,
+    ensure_all_sources_have_names,
+    location_to_string,
+    prune_unknown_kwargs,
+)
+
+
+@dataclass
+class BaseCompiler:
+    postProcessorClass: Type = PostProcessor
+    featureCompilerClass: Optional[Type] = None
+    featureWriters: list = None
+    filters: list = None
+    glyphOrder: list = None
+    useProductionNames: bool = None
+    removeOverlaps: bool = False
+    overlapsBackend: str = None
+    inplace: bool = False
+    layerName: str = None
+    skipExportGlyphs: bool = None
+    debugFeatureFile: str = None
+    notdefGlyph: str = None
+    colrLayerReuse: bool = True
+    colrAutoClipBoxes: bool = True
+    colrClipBoxQuantization: Callable[[object], int] = colrClipBoxQuantization
+    feaIncludeDir: str = None
+    skipFeatureCompilation: bool = False
+    flattenComponents: bool = None
+    _tables: list = None
+
+    def __post_init__(self):
+        self.logger = logging.getLogger(self.__class__.__name__)
+        self.timer = Timer(logging.getLogger("ufo2ft.timer"), level=logging.DEBUG)
+
+    def compile(self, ufo):
+        glyphSet = self.preprocess(ufo)
+        font = self.compileOutlines(ufo, glyphSet)
+        if self.layerName is None and not self.skipFeatureCompilation:
+            self.compileFeatures(ufo, font, glyphSet=glyphSet)
+        return self.postprocess(font, ufo, glyphSet)
+
+    def preprocess(self, ufo_or_ufos):
+        self.logger.info("Pre-processing glyphs")
+        if self.skipExportGlyphs is None:
+            if isinstance(ufo_or_ufos, (list, tuple)):
+                self.skipExportGlyphs = set()
+                for ufo in ufo_or_ufos:
+                    self.skipExportGlyphs.update(
+                        ufo.lib.get("public.skipExportGlyphs", [])
+                    )
+            else:
+                self.skipExportGlyphs = ufo_or_ufos.lib.get(
+                    "public.skipExportGlyphs", []
+                )
+
+        # Preprocessors expect this parameter under a different name.
+        if hasattr(self, "cubicConversionError"):
+            self.conversionError = self.cubicConversionError
```

Comment:
can we not modify self.__dict__ and instead create a copy of this dict to pass on as kwargs to prepropcessor?